src/simulation/ApplyLoad.cpp

```diff
@@ -0,0 +1,266 @@
+    upgrade.push_back(UpgradeType{v.begin(), v.end()});
+
+    closeLedger({invokeTx.second}, upgrade);
+}
+
+void
+ApplyLoad::setupLoadContracts()
+{
+    auto wasm = rust_bridge::get_test_wasm_loadgen();
+    xdr::opaque_vec<> wasmBytes;
+    wasmBytes.assign(wasm.data.begin(), wasm.data.end());
+
+    LedgerKey contractCodeLedgerKey;
+    contractCodeLedgerKey.type(CONTRACT_CODE);
+    contractCodeLedgerKey.contractCode().hash = sha256(wasmBytes);
+
+    mLoadCodeKey = contractCodeLedgerKey;
+
+    auto const& lm = mApp.getLedgerManager();
+    auto uploadTx = mTxGenerator.createUploadWasmTransaction(
+        lm.getLastClosedLedgerNum() + 1, 0, wasmBytes, contractCodeLedgerKey,
+        std::nullopt);
+
+    closeLedger({uploadTx.second});
+
+    for (auto kvp : mTxGenerator.getAccounts())
+    {
+        auto salt = sha256("Load contract " + std::to_string(kvp.first));
+
+        auto createTx = mTxGenerator.createContractTransaction(
+            lm.getLastClosedLedgerNum() + 1, 0, contractCodeLedgerKey,
+            wasmBytes.size() + 160, salt, std::nullopt);
+        closeLedger({createTx.second});
+
+        auto instanceKey =
+            createTx.second->sorobanResources().footprint.readWrite.back();
+
+        TxGenerator::ContractInstance instance;
+        instance.readOnlyKeys.emplace_back(mLoadCodeKey);
+        instance.readOnlyKeys.emplace_back(instanceKey);
+        instance.contractID = instanceKey.contractData().contract;
+        mLoadInstances.emplace(kvp.first, instance);
+    }
+}
+
+void
+ApplyLoad::benchmark()
+{
+    auto& lm = mApp.getLedgerManager();
+    std::vector<TransactionFrameBasePtr> txs;
+
+    auto resources = multiplyByDouble(lm.maxLedgerResources(true),
+                                      TRANSACTION_QUEUE_SIZE_MULTIPLIER);
+
+    auto const& accounts = mTxGenerator.getAccounts();
+    std::vector<uint64_t> shuffledAccounts(accounts.size());
+    std::iota(shuffledAccounts.begin(), shuffledAccounts.end(), 0);
+    stellar::shuffle(std::begin(shuffledAccounts), std::end(shuffledAccounts),
+                     gRandomEngine);
+
+    for (auto accountIndex : shuffledAccounts)
+    {
+        auto it = accounts.find(accountIndex);
+        releaseAssert(it != accounts.end());
+
+        auto instanceIter = mLoadInstances.find(it->first);
+        releaseAssert(instanceIter != mLoadInstances.end());
+        auto const& instance = instanceIter->second;
+        auto tx = mTxGenerator.invokeSorobanLoadTransaction(
+            lm.getLastClosedLedgerNum() + 1, it->first, instance,
+            rust_bridge::get_write_bytes().data.size() + 160, std::nullopt);
+
+        if (!anyGreater(tx.second->getResources(false), resources))
+        {
+            resources -= tx.second->getResources(false);
+        }
+        else
+        {
+            for (size_t i = static_cast<size_t>(Resource::Type::OPERATIONS);
+                 i <= static_cast<size_t>(Resource::Type::WRITE_LEDGER_ENTRIES);
```

Comment:
Why do we not iterate all the resource types from 0 to `resources.size()`?

Reply:
Ah I didn't realize it had a size method. Fixed.